src/xamarin/sk_manageddrawable.cpp

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2015 Xamarin Inc.
+ * Copyright 2017 Microsoft Corporation. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+
+#include "SkManagedDrawable.h"
+
+#include "sk_manageddrawable.h"
+#include "sk_picture.h"
+#include "sk_types_priv.h"
+
+/* This is a C wrapper around the C++ class, so that it is possible
+   to access the C++ from C#.                                        */
+   
+static sk_manageddrawable_draw_delegate               gDraw;
+static sk_manageddrawable_getBounds_delegate          gGetBounds;
+static sk_manageddrawable_newPictureSnapshot_delegate gNewPictureSnapshot;
+
+static inline SkManagedDrawable* AsManagedDrawable(sk_manageddrawable_t* cdrawable) 
+{
+    return reinterpret_cast<SkManagedDrawable*>(cdrawable);
+}
+
+static inline sk_manageddrawable_t* ToManagedDrawable(SkManagedDrawable* drawable) 
+{
+    return reinterpret_cast<sk_manageddrawable_t*>(drawable);
+}
+
+static inline const sk_manageddrawable_t* ToManagedDrawable(const SkManagedDrawable* drawable) 
+{
+    return reinterpret_cast<const sk_manageddrawable_t*>(drawable);
+}
+
+void dDraw(SkManagedDrawable* managedDrawable, SkCanvas* canvas) 
+{
+    gDraw(ToManagedDrawable(managedDrawable), ToCanvas(canvas));
+}
+
+void dGetBounds(SkManagedDrawable* managedDrawable, SkRect* rect) 
+{
+    gGetBounds(ToManagedDrawable(managedDrawable), ToRect(rect));
+}
+
+SkPicture* dNewPictureSnapshot(SkManagedDrawable* managedDrawable) 
+{
+    return AsPicture(gNewPictureSnapshot(ToManagedDrawable(managedDrawable)));
+}
+
+sk_manageddrawable_t* sk_manageddrawable_new() { 
+	return ToManagedDrawable(new SkManagedDrawable());
+}
+
+void sk_manageddrawable_destroy(sk_manageddrawable_t* drawable)
+{ 
+    delete AsManagedDrawable(drawable);
```

Comment:
The constructor and destructor may have to change... I was having a look at the code and I think this may cause issues because this object is reference counted in skia (`SkDrawable : SkFlattenable : SkRefCnt`):

```csharp
var myDrawable = new MyDrawable(); // ref count of 1
anotherObject.SetDrawable(myDrawable); // ref count = 2
myDrawable.Dispose(); // hard delete but anotherObject is still referencing it
```

I didn't notice it at first, but all the object that I have explicitly `new`-ed up have not been reference counted, and were never potentially held onto for longer than the called method. Howerver, `SkDrawable` is. For those cases, you need to do this:

```suggestion
    SkSafeUnref(AsManagedDrawable(drawable));
```

In 2 cases in `sk_typeface.h`, i am calling `->unref()` directly, but that is not safe because the object may be null or disposed alread. The code for `SkSafeUnref` is:

```cpp
template <typename T> static inline void SkSafeUnref(T* obj) {
    if (obj) {
        obj->unref();
    }
}
```